clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java

```diff
@@ -0,0 +1,354 @@
+            }
+            if (element.equals(newElement)) {
+                return INVALID_INDEX;
+            }
+            slot = (slot + 1) % addElements.length;
+        }
+        throw new RuntimeException("Not enough hash table slots to add a new element.");
+    }
+
+    private void changeCapacity(int newCapacity) {
+        Element[] newElements = new Element[newCapacity];
+        HeadElement newHead = new HeadElement();
+        int oldSize = size;
+        for (Iterator<E> iter = iterator(); iter.hasNext(); ) {
+            Element element = iter.next();
+            iter.remove();
+            int newSlot = addInternal(element, newElements);
+            addToListTail(newHead, newElements, newSlot);
+        }
+        this.elements = newElements;
+        this.head = newHead;
+        this.size = oldSize;
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        E example = null;
+        try {
+            example = (E) o;
+        } catch (ClassCastException e) {
+            return false;
+        }
+        int slot = findIndex(example);
+        if (slot == INVALID_INDEX) {
+            return false;
+        }
+        size--;
+        removeFromList(head, elements, slot);
+        slot = (slot + 1) % elements.length;
+
+        // Find the next empty slot
+        int endSlot = slot;
+        for (int seen = 0; seen < elements.length; seen++) {
+            Element element = elements[endSlot];
+            if (element == null) {
+                break;
+            }
+            endSlot = (endSlot + 1) % elements.length;
+        }
+
+        // We must preserve the denseness invariant.  The denseness invariant says that
+        // any element is either in the slot indicated by its hash code, or a slot which
+        // is not separated from that slot by any nulls.
+        // Reseat all elements in between the deleted element and the next empty slot.
+        while (slot != endSlot) {
+            reseat(slot);
+            slot = (slot + 1) % elements.length;
+        }
+        return true;
+    }
+
+    private void reseat(int prevSlot) {
+        Element element = elements[prevSlot];
+        int newSlot = slot(elements, element);
+        for (int seen = 0; seen < elements.length; seen++) {
+            Element e = elements[newSlot];
+            if ((e == null) || (e == element)) {
+                break;
+            }
+            newSlot = (newSlot + 1) % elements.length;
+        }
+        if (newSlot == prevSlot) {
+            return;
+        }
+        Element prev = indexToElement(head, elements, element.prev());
+        prev.setNext(newSlot);
+        Element next = indexToElement(head, elements, element.next());
+        next.setPrev(newSlot);
+        elements[prevSlot] = null;
+        elements[newSlot] = element;
+    }
+
+    @Override
+    public void clear() {
+        reset(elements.length);
+    }
+
+    public ImplicitLinkedHashSet() {
+        this(10);
```

Comment:
Note that `Hashtable` uses 11 (a prime number) as the default.

Reply:
I will change this to 5, so that we also get 11 as the default number of slots.